src/zenml/zen_server/exceptions.py

```diff
@@ -0,0 +1,175 @@
+#  Copyright (c) ZenML GmbH 2023. All Rights Reserved.
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at:
+#
+#       https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+#  or implied. See the License for the specific language governing
+#  permissions and limitations under the License.
+"""REST API exception handling."""
+
+from typing import TYPE_CHECKING, Any, List, Optional, Tuple, Type
+
+import requests
+from pydantic import BaseModel
+from pydantic import ValidationError as PydanticValidationError
+
+from zenml.exceptions import (
+    AuthorizationException,
+    DoesNotExistException,
+    DuplicateRunNameError,
+    EntityExistsError,
+    IllegalOperationError,
+    SecretExistsError,
+    StackComponentExistsError,
+    StackExistsError,
+    ValidationError,
+    ZenKeyError,
+)
+
+if TYPE_CHECKING:
+    from fastapi import HTTPException
+
+
+class ErrorModel(BaseModel):
+    """Base class for error responses."""
+
+    detail: Any
+
+
+error_response = dict(model=ErrorModel)
+
+# Map exceptions to HTTP status codes. This is used in two ways:
+# 1. When an exception is raised in the server, the exception is converted to a
+#   error response with the appropriate status code and with the exception class
+#   name and arguments as the response body.
+# 2. When HTTP error response is received by the client, the error response is
+#  converted back to an exception based on the status code and the exception
+#  class name and arguments are extracted from the response body.
+#
+# NOTE: The order of the exceptions is important. The first matching
+# exception will be used, so the more specific exceptions should be
+# listed first. For error responses that cannot be unpacked into an
+# exception, the last (the more generic) exception in the list corresponding
+# to the status code will be used.
+REST_API_EXCEPTIONS: List[Tuple[Type[Exception], int]] = [
```

Comment:
Using a dictionary for the "exception to error code" specification wouldn't work, for one main reason: the lookup operation used to determine which error code to use for an exception is not equality (`==`) but `isinstance` instead. This allows a progression from more specific to less specific exception types, same you would do in a try-catch block with multiple errors. This also allows us to convert any future exception into an error code as long as it inherits from one of these as a base exception without having to add it explicitly to this list.

I also thought about splitting it into two maps, but it would duplicate a lot of information and make it more error-prone to maintain it in the future.